diff -ur pbench-sysstat-11.1.2/Makefile.in pbench-sysstat-11.1.2-at/Makefile.in
--- pbench-sysstat-11.1.2/Makefile.in	2014-10-19 09:39:41.000000000 -0400
+++ pbench-sysstat-11.1.2-at/Makefile.in	2014-10-20 16:30:46.000000000 -0400
@@ -330,7 +330,7 @@
 	done
 endif
 
-install_base: all sa1 sa2 sysstat.sysconfig install_man install_nls \
+install_base: all sa1 sa2 install_man install_nls \
 	contrib/isag/isag
 	mkdir -p $(DESTDIR)$(SA_LIB_DIR)
 	mkdir -p $(DESTDIR)$(SA_DIR)
@@ -355,8 +355,6 @@
 ifeq ($(INSTALL_ISAG),y)
 	$(INSTALL_BIN) contrib/isag/isag $(DESTDIR)$(BIN_DIR)
 endif
-	$(INSTALL_DATA) sysstat.ioconf $(DESTDIR)$(SYSCONFIG_DIR)
-	$(INSTALL_DATA) sysstat.sysconfig $(DESTDIR)$(SYSCONFIG_DIR)/sysstat
 ifeq ($(INSTALL_DOC),y)
 	$(INSTALL_DATA) CHANGES $(DESTDIR)$(DOC_DIR)
 	$(INSTALL_DATA) COPYING $(DESTDIR)$(DOC_DIR)
